utils/staking-miner/src/monitor.rs

```diff
@@ -64,120 +68,245 @@ async fn ensure_no_previous_solution<
 }
 
 macro_rules! monitor_cmd_for { ($runtime:tt) => { paste::paste! {
+
 	/// The monitor command.
 	pub(crate) async fn [<monitor_cmd_ $runtime>](
-		client: &WsClient,
+		client: WsClient,
 		shared: SharedConfig,
 		config: MonitorConfig,
 		signer: Signer,
 	) -> Result<(), Error<$crate::[<$runtime _runtime_exports>]::Runtime>> {
 		use $crate::[<$runtime _runtime_exports>]::*;
+		type StakingMinerError = Error<$crate::[<$runtime _runtime_exports>]::Runtime>;
 
 		let (sub, unsub) = if config.listen == "head" {
 			("chain_subscribeNewHeads", "chain_unsubscribeNewHeads")
 		} else {
 			("chain_subscribeFinalizedHeads", "chain_unsubscribeFinalizedHeads")
 		};
 
+		let mut subscription: Subscription<Header> = client.subscribe(&sub, None, &unsub).await?;
+
+		let client = Arc::new(client);
+		let (tx, mut rx) = mpsc::unbounded_channel::<StakingMinerError>();
+
 		loop {
-			log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", sub, unsub);
-			let mut subscription: Subscription<Header> = client
-				.subscribe(&sub, None, &unsub)
-				.await
-				.unwrap();
-
-			while let Some(now) = subscription.next().await? {
-				let hash = now.hash();
-				log::trace!(target: LOG_TARGET, "new event at #{:?} ({:?})", now.number, hash);
-
-				// if the runtime version has changed, terminate.
-				crate::check_versions::<Runtime>(client).await?;
-
-				// we prefer doing this check before fetching anything into a remote-ext.
-				if ensure_signed_phase::<Runtime, Block>(client, hash).await.is_err() {
-					log::debug!(target: LOG_TARGET, "phase closed, not interested in this block at all.");
-					continue;
-				};
+			let at = tokio::select! {
+				maybe_rp = subscription.next() => {
+					match maybe_rp {
+						Some(Ok(r)) => r,
+						// Custom `jsonrpsee` message; should not occur.
+						Some(Err(RpcError::SubscriptionClosed(reason))) => {
```

Comment:
For the `Some(Err(_))` cases why not an error but still retry with `subscription = client.subscribe(&sub, None, &unsub).await?;`?

Reply:
(Not saying we should change this, but just want to understand the rational)

Reply:
good catch, yupp I agree that we should create a new subscription if this occurs.

currently this is unreachable anyways because it's a custom message we have introduced in jsonrpsee once the server closes down subscription. I don't think jsonrpc has this FWIW.